implementation/src/main/java/io/smallrye/config/RelocateConfigSourceInterceptor.java

```diff
@@ -0,0 +1,18 @@
+package io.smallrye.config;
+
+import javax.annotation.Priority;
+
+@Priority(200)
+public class RelocateConfigSourceInterceptor implements ConfigSourceInterceptor {
+    private static final String RELOCATE_PREFIX = "%relocate.";
```

Comment:
What would be the use case of this interceptor as written?

Reply:
Yes, added this for us to discuss :)

The idea was to implement relocation similar to what Quarkus did with the profiles, but with the relocation prefix in the value instead.

**Use Case**: I want to relocate / fallback to a different configuration name in case I change my configuration name or another configuration name is introduced with the same business behaviour (#268). Also, I want to be able to prioritise such relocations / fallbacks.

**Possible Implementation**:
Relocation happens when the configuration value has the prefix `%relocate.` followed by the new key name.

**Sample 1: (simple relocate)**
```properties
smallrye.jwt.token.header=%relocate.mp.jwt.token.header
mp.jwt.token.header=Authorization
```

I search a configuration for `smallrye.jwt.token.header` or `mp.jwt.token.header` and get `Authorization` in both.

**Sample 2: (specific config with higher priority)**
```properties priority 1000
smallrye.jwt.token.header=Authorization
```
```properties priority 900
smallrye.jwt.token.header=%relocate.mp.jwt.token.header
mp.jwt.token.header=Cookie
```

I search a configuration for `smallrye.jwt.token.header` and get `Authorization`.

I'm happy to hear other ideas. This seems to be a clean approach, because it will give you control at the ConfigSource plus priority on what you want to relocate and it will be very explicit. Now, if we have a complete namespace change and have to do this for a bunch of keys, we can expand this one and add a regex style configuration.

Reply:
First I'll note (for my own future reference) that the relationship with Quarkus profiles is purely related to the usage of `%` (this change does not introduce profiles per se).

> **Possible Implementation**:
> Relocation happens when the configuration value has the prefix `%relocate.` followed by the new key name.

This implementation is functionally identical to having a value which is an expression referencing the new key name.  So this is really just a second way to accomplish this exact thing; I would argue we should stick with just the one implementation.  So (for example) sample 1 could be exactly accomplished by `smallrye.jwt.token.header=${mp.jwt.token.header:}` without introducing a second mechanism.

Going back to the use case:

> **Use Case**: I want to relocate / fallback to a different configuration name in case I change my configuration name or another configuration name is introduced with the same business behaviour (#268). Also, I want to be able to prioritise such relocations / fallbacks.

This can be done as well using a more blanket interception.  I think this use case broadly falls under one of two categories:

* A configuration namespace has relocated to a new namespace
* A single configuration property is been moved to a new location

For the first case, I think a prefix-matching property mapper is effective.  By using interceptors to access properties at both the old and new locations, any prioritization rule that makes sense can be applied; furthermore, it is very memory efficient, not requiring the retention of the complete set of keys with placeholder values.

For the second case, a plain string to string mapping would be effective, similarly via interception.

As a final note: Quarkus has explicitly avoided supporting any kind of relocation of properties, generally speaking, since it is presumed that it would cause some usability difficulties.

Reply:
> First I'll note (for my own future reference) that the relationship with Quarkus profiles is purely related to the usage of `%` (this change does not introduce profiles per se).
> 

I've just said that the notion was inspired by Quarkus profile, but in no way wants to reimplement Quarkus notion of profile.

> > **Possible Implementation**:
> > Relocation happens when the configuration value has the prefix `%relocate.` followed by the new key name.
> 
> This implementation is functionally identical to having a value which is an expression referencing the new key name. So this is really just a second way to accomplish this exact thing; I would argue we should stick with just the one implementation. So (for example) sample 1 could be exactly accomplished by `smallrye.jwt.token.header=${mp.jwt.token.header:}` without introducing a second mechanism.

Well, that was my first proposal, and I thought you were worried about the expressions objects, so I took a more direct approach with no expression involved to retrieve the config directly. If it is acceptable to use expressions for this use case, I'm happy to implement it that way.

> 
> Going back to the use case:
> 
> > **Use Case**: I want to relocate / fallback to a different configuration name in case I change my configuration name or another configuration name is introduced with the same business behaviour (#268). Also, I want to be able to prioritise such relocations / fallbacks.
> 
> This can be done as well using a more blanket interception. I think this use case broadly falls under one of two categories:
> 
> * A configuration namespace has relocated to a new namespace
> * A single configuration property is been moved to a new location
> 
> For the first case, I think a prefix-matching property mapper is effective. By using interceptors to access properties at both the old and new locations, any prioritization rule that makes sense can be applied; furthermore, it is very memory efficient, not requiring the retention of the complete set of keys with placeholder values.
> 
> For the second case, a plain string to string mapping would be effective, similarly via interception.

Ok. But where do you suggest to write these mappings / rules? For me, the obvious choice is to add them as some sort of config, and these requires to choose some kind of discriminator so we don't clash with configs already into place. Or do you think we should not provide a basic implementation to serve the simple use cases and each project / app should implement their own?

> 
> As a final note: Quarkus has explicitly avoided supporting any kind of relocation of properties, generally speaking, since it is presumed that it would cause some usability difficulties.

I understand. But I think this is a clear use case and we are experiencing it now with SmallRye JWT. It is very likely that is going to happen a few more times in the future for other specs, while we keep innovating at the implementations and pushing stuff for the API's. Do you think we should drop this and let projects deal with this as they see fit?

Reply:
> > > **Possible Implementation**:
> > > Relocation happens when the configuration value has the prefix `%relocate.` followed by the new key name.
> > 
> > 
> > This implementation is functionally identical to having a value which is an expression referencing the new key name. So this is really just a second way to accomplish this exact thing; I would argue we should stick with just the one implementation. So (for example) sample 1 could be exactly accomplished by `smallrye.jwt.token.header=${mp.jwt.token.header:}` without introducing a second mechanism.
> 
> Well, that was my first proposal, and I thought you were worried about the expressions objects, so I took a more direct approach with no expression involved to retrieve the config directly. If it is acceptable to use expressions for this use case, I'm happy to implement it that way.

I am still concerned about using expression objects for this; this proposed alternative avoids creating expression objects but still requires a multitude of objects on a similar order of magnitude, particularly in the whole-namespace scenario.  Bulk mapping is substantially more memory-efficient.

> > Going back to the use case:
> > > **Use Case**: I want to relocate / fallback to a different configuration name in case I change my configuration name or another configuration name is introduced with the same business behaviour (#268). Also, I want to be able to prioritise such relocations / fallbacks.
> > 
> > 
> > This can be done as well using a more blanket interception. I think this use case broadly falls under one of two categories:
> > 
> > * A configuration namespace has relocated to a new namespace
> > * A single configuration property is been moved to a new location
> > 
> > For the first case, I think a prefix-matching property mapper is effective. By using interceptors to access properties at both the old and new locations, any prioritization rule that makes sense can be applied; furthermore, it is very memory efficient, not requiring the retention of the complete set of keys with placeholder values.
> > For the second case, a plain string to string mapping would be effective, similarly via interception.
> 
> Ok. But where do you suggest to write these mappings / rules? For me, the obvious choice is to add them as some sort of config, and these requires to choose some kind of discriminator so we don't clash with configs already into place. Or do you think we should not provide a basic implementation to serve the simple use cases and each project / app should implement their own?

I don't think config is the right place for such rules.  The remappings in both scenarios above are driven by the configuration consumer, not the end user.  Putting mapping rules in the configuration puts it into the end user's arena.

It makes more sense to me for the configuration consumer (i.e. the framework) to have some way to express that the remappings should be done.  Publishing an interceptor is a convenient way to achieve this, in addition to enabling the more efficient possible implementations I mentioned.

SmallRye Config could enable these cases by creating interceptor classes which implement this functionality.  With what we have today, users could extend these classes with variations that populate the mappings as needed upon construction; or, as a more elegant solution, we could introduce a `ConfigSourceInterceptorProvider` in the same spirit as `ConfigSourceProvider` which can construct and return configured instances of these classes.

> > As a final note: Quarkus has explicitly avoided supporting any kind of relocation of properties, generally speaking, since it is presumed that it would cause some usability difficulties.
> 
> I understand. But I think this is a clear use case and we are experiencing it now with SmallRye JWT. It is very likely that is going to happen a few more times in the future for other specs, while we keep innovating at the implementations and pushing stuff for the API's. Do you think we should drop this and let projects deal with this as they see fit?

I think the use case is valid.  I do think that the projects need to have firsthand control over what gets mapped, since that's strictly in their domain.  Providing convenience interceptors could be an effective way to achieve this without polluting the API.

Reply:
> I am still concerned about using expression objects for this; this proposed alternative avoids creating expression objects but still requires a multitude of objects on a similar order of magnitude, particularly in the whole-namespace scenario. Bulk mapping is substantially more memory-efficient.
> 

This was to cover simple cases. It was not the intent to support the whole namespace scenario relocation. That will need to be implemented as you said with some sort of prefix match provider. 

There is another detail that I haven't mentioned. Again, in the case of SmallRye JWT, there are properties in the `smallrye.jwt` namespace that do not have corresponding equivalents in the `mp.jwt` namespace. Either they don't make sense at the API level, or not yet field proven to move the the MP API. In this situation, we cannot blindly remap all the properties in one namespace into the other.

> 
> I don't think config is the right place for such rules. The remappings in both scenarios above are driven by the configuration consumer, not the end user. Putting mapping rules in the configuration puts it into the end user's arena.
> 
> It makes more sense to me for the configuration consumer (i.e. the framework) to have some way to express that the remappings should be done. Publishing an interceptor is a convenient way to achieve this, in addition to enabling the more efficient possible implementations I mentioned.
> 
> SmallRye Config could enable these cases by creating interceptor classes which implement this functionality. With what we have today, users could extend these classes with variations that populate the mappings as needed upon construction; or, as a more elegant solution, we could introduce a `ConfigSourceInterceptorProvider` in the same spirit as `ConfigSourceProvider` which can construct and return configured instances of these classes.

That is fine. I was planning to introduce the `Provider` to the interceptors as well. But I also wanted to have a simple interceptor that you could use with minimal effort. In this cases, what do you think about introducing a `relocate.properties` for a simple one to one mapping case?

Also,  I think end users should be able to use this if they want to use it.

> 
> I think the use case is valid. I do think that the projects need to have firsthand control over what gets mapped, since that's strictly in their domain. Providing convenience interceptors could be an effective way to achieve this without polluting the API.

Ok. That is what I want to achieve as well. Thanks!

Reply:
> > I am still concerned about using expression objects for this; this proposed alternative avoids creating expression objects but still requires a multitude of objects on a similar order of magnitude, particularly in the whole-namespace scenario. Bulk mapping is substantially more memory-efficient.
> 
> This was to cover simple cases. It was not the intent to support the whole namespace scenario relocation. That will need to be implemented as you said with some sort of prefix match provider.
> 
> There is another detail that I haven't mentioned. Again, in the case of SmallRye JWT, there are properties in the `smallrye.jwt` namespace that do not have corresponding equivalents in the `mp.jwt` namespace. Either they don't make sense at the API level, or not yet field proven to move the the MP API. In this situation, we cannot blindly remap all the properties in one namespace into the other.

All the more reason to use an interceptor base class for this purpose: it can implement arbitrarily complex remapping decisions with minimal cost.  Example pseudocode:

```java
if (name.startsWith("prefix1.") && ! name.startsWith("prefix1.subprefix2") && name.endsWith(".remapMe")) { ... }
```

This example approach allocates zero additional objects per name lookup and yet can perform complex decisions.

> > I don't think config is the right place for such rules. The remappings in both scenarios above are driven by the configuration consumer, not the end user. Putting mapping rules in the configuration puts it into the end user's arena.
> > It makes more sense to me for the configuration consumer (i.e. the framework) to have some way to express that the remappings should be done. Publishing an interceptor is a convenient way to achieve this, in addition to enabling the more efficient possible implementations I mentioned.
> > SmallRye Config could enable these cases by creating interceptor classes which implement this functionality. With what we have today, users could extend these classes with variations that populate the mappings as needed upon construction; or, as a more elegant solution, we could introduce a `ConfigSourceInterceptorProvider` in the same spirit as `ConfigSourceProvider` which can construct and return configured instances of these classes.
> 
> That is fine. I was planning to introduce the `Provider` to the interceptors as well. But I also wanted to have a simple interceptor that you could use with minimal effort. In this cases, what do you think about introducing a `relocate.properties` for a simple one to one mapping case?

Again I don't agree with doing it this way: the cost is higher and it's not clear what the benefit is.  We have just introduced a flexible interception framework.  It is best to put this into the users' hands.

> Also, I think end users should be able to use this if they want to use it.

Ultimately you have to decide what use case is being met.  So far everything discussed relates to framework providers, not to end users.  But even in the end user case, they can always produce their own interceptor.

Reply:
> Again I don't agree with doing it this way: the cost is higher and it's not clear what the benefit is. We have just introduced a flexible interception framework. It is best to put this into the users' hands.

It would still be implemented in the interceptor, but the idea was to read the simple mappings from that file, so you didn't have to implement your own interceptor or Provider to init a possible base class.

Goal was to not have to implement anything on SmallRye JWT side (since this is a simple use case). If things become less evident, that sure, we implement the interceptor over there with the expected rules.

Do you think we should provide a way with minimal configuration to handle simple cases? If yes, how does the consumer configures the mappings / rules?

Reply:
Once you introduce a file-based API, you've adopted a permanent maintenance burden which is beyond a simple API.  The file needs a syntax, and the syntax needs to be concise and also needs to meet the use case(s), which need to be enumerated.  And when new cases arise, you hope that the syntax can be extended to meet them but that's not a given at all.

An interceptor class could trivially be implemented which accepts a name-mapping `Function` which is completely open-ended and flexible, and also easy to use.  But even then you still have a semantic decision relating to whether the configuration is absolutely relocated or whether the original location is also allowed to have a value.  So that again comes down to use cases.

Perhaps a different approach would be, examine the use cases, and see how difficult it is to meet each use case with *just* an interceptor.  If the resultant interceptor is reasonably simple, then I'd say it's just not worth adding more maintenance surface area.

There needs to be sufficient value provided to justify the cost of the feature.  If the problem is trivial to solve with an interceptor then to me the value is not sufficient.

Reply:
That is fine, I'll add the Provider support so it would be easier for consumers to create their interceptors and provide a base mapper interceptor as well.

I think also my main concern is to provide something that users can see how it works (either by a base class and / or documentation) so they know that they can do it in that way.

Reply:
Maybe we need some example subprojects?

Reply:
Yes, that already come to my mind multiple times.

The issue is where? Most consumers are not even aware of SmallRye (they know Quarkus or Wildfly). So if you add the samples here, they have a very low chance of being found (you have to promote them, blog them, etc). 

I guess that if we link them from either Wildfly / Quarkus to here it helps. 

What do you think?

Reply:
It sounds like examples are nice to have but not likely to be worth the investment (at present).  Maybe that will change in the future.

Reply:
If we do the base class, plus add it to the documentation, at least that will be something we can point people to.